internal/pkg/aws/s3/s3.go

```diff
@@ -194,68 +195,99 @@ func FormatARN(partition, location string) string {
 	return fmt.Sprintf("arn:%s:s3:::%s", partition, location)
 }
 
-func (s *S3) bucketExists(bucket string) (bool, error) {
-	input := &s3.HeadBucketInput{
-		Bucket: aws.String(bucket),
+// BucketTree creates an ASCII tree representing the folder structure of a bucket's objects.
+func (s *S3) BucketTree(bucket string) (string, error) {
+	outputs, err := s.listObjects(bucket, "/")
+	if err != nil {
+		return "", err
 	}
-	_, err := s.s3Client.HeadBucket(input)
+	if outputs == nil {
+		return "", nil
+	}
+	var contents []*s3.Object
+	var prefixes []*s3.CommonPrefix
+	for _, output := range outputs {
+		contents = append(contents, output.Contents...)
+		prefixes = append(prefixes, output.CommonPrefixes...)
+	}
+	tree := treeprint.New()
+	// Add top-level files.
+	for _, object := range contents {
+		tree.AddNode(aws.StringValue(object.Key))
+	}
+	// Recursively add folders and their children.
+	if err := s.addNodes(tree, prefixes, bucket); err != nil {
+		return "", err
+	}
+	return tree.String(), nil
+}
+
+// BucketSizeAndCount returns the total size and number of objects in an S3 bucket.
+func (s *S3) BucketSizeAndCount(bucket string) (string, int, error) {
+	outputs, err := s.listObjects(bucket, "")
 	if err != nil {
-		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == errCodeNotFound {
-			return false, nil
+		return "", 0, err
+	}
+	if outputs == nil {
+		return "", 0, nil
+	}
+	var size int64
+	var number int
+	for _, output := range outputs {
+		for _, object := range output.Contents {
+			size = size + aws.Int64Value(object.Size)
+			number = number + 1
 		}
-		return false, err
 	}
-
-	return true, nil
+	return humanize.Bytes(uint64(size)), number, nil
 }
 
-// GetBucketTree retrieves the objects in an S3 bucket and creates an ASCII tree representing their folder structure.
-func (s *S3) GetBucketTree(bucket string) (string, error) {
+func (s *S3) listObjects(bucket, delimiter string) ([]s3.ListObjectsV2Output, error) {
 	exists, err := s.bucketExists(bucket)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 	if !exists {
-		return "", nil
+		return nil, nil
 	}
-
-	var contents []*s3.Object
-	var prefixes []*s3.CommonPrefix
+	var outputs []s3.ListObjectsV2Output
 	listResp := &s3.ListObjectsV2Output{}
 	for {
 		listParams := &s3.ListObjectsV2Input{
 			Bucket:            aws.String(bucket),
-			Delimiter:         aws.String(slashDelimiter),
+			Delimiter:         aws.String(delimiter),
 			ContinuationToken: listResp.NextContinuationToken,
 		}
 		listResp, err = s.s3Client.ListObjectsV2(listParams)
 		if err != nil {
-			return "", fmt.Errorf("list objects for bucket %s: %w", bucket, err)
+			return nil, fmt.Errorf("list objects for bucket %s: %w", bucket, err)
 		}
-		contents = append(contents, listResp.Contents...)
-		prefixes = append(prefixes, listResp.CommonPrefixes...)
+		outputs = append(outputs, *listResp)
```

Comment:
i wonder if it makes sense to merge the output into a single `s3.ListObjectsV2Output` instead? then the other functions only have to loop once?

Reply:
That would be nice, but we have to keep the prefixes and contents separate because in the recursive call we use each prefix as a delimiter.